core/src/main/java/io/questdb/griffin/engine/functions/str/SplitPartFunctionFactory.java

```diff
@@ -0,0 +1,210 @@
+        final Function strFunc = args.getQuick(0);
+        final Function delimiterFunc = args.getQuick(1);
+        final Function indexFunc = args.getQuick(2);
+        if (indexFunc.isConstant()) {
+            int index = indexFunc.getInt(null);
+            if (index == Numbers.INT_NaN) {
+                return StrConstant.NULL;
+            } else if (index == 0) {
+                return StrConstant.EMPTY;
+            } else {
+                return new SplitPartConstIndexFunction(strFunc, delimiterFunc, index);
+            }
+        }
+        return new SplitPartFunction(strFunc, delimiterFunc, indexFunc);
+    }
+
+    private static class SplitPartFunction extends AbstractSplitPartConstIndexFunction implements TernaryFunction {
+        private final Function strFunc;
+        private final Function delimiterFunc;
+        private final Function indexFunc;
+
+        public SplitPartFunction(Function strFunc, Function delimiterFunc, Function indexFunc) {
+            super(strFunc, delimiterFunc);
+            this.strFunc = strFunc;
+            this.delimiterFunc = delimiterFunc;
+            this.indexFunc = indexFunc;
+        }
+
+        @Override
+        int getIndex(Record rec) {
+            return indexFunc.getInt(rec);
+        }
+
+        @Override
+        public Function getLeft() {
+            return strFunc;
+        }
+
+        @Override
+        public Function getCenter() {
+            return delimiterFunc;
+        }
+
+        @Override
+        public Function getRight() {
+            return indexFunc;
+        }
+    }
+
+    private static class SplitPartConstIndexFunction extends AbstractSplitPartConstIndexFunction implements BinaryFunction {
+        private final Function strFunc;
+        private final Function delimiterFunc;
+        private final int index;
+
+        public SplitPartConstIndexFunction(Function strFunc, Function delimiterFunc, int index) {
+            super(strFunc, delimiterFunc);
+            this.strFunc = strFunc;
+            this.delimiterFunc = delimiterFunc;
+            this.index = index;
+        }
+
+        @Override
+        int getIndex(Record rec) {
+            return index;
+        }
+
+        @Override
+        public Function getLeft() {
+            return strFunc;
+        }
+
+        @Override
+        public Function getRight() {
+            return delimiterFunc;
+        }
+    }
+
+    private static abstract class AbstractSplitPartConstIndexFunction extends StrFunction {
+        private final StringSink sink = new StringSink();
+        private final StringSink sinkB = new StringSink();
+
+        private final Function strFunc;
+        private final Function delimiterFunc;
+
+        public AbstractSplitPartConstIndexFunction(Function strFunc, Function delimiterFunc) {
+            this.strFunc = strFunc;
+            this.delimiterFunc = delimiterFunc;
+        }
+
+        abstract int getIndex(Record rec);
```

Comment:
I see a lot of code duplication between the function implementation that takes in a `Function` vs a constant value. To avoid that, I created an abstract class that has an abstract method that returns the value.